rest/rest-resources/src/test/java/org/apache/brooklyn/rest/resources/BundleAndTypeResourcesTest.java

```diff
@@ -1095,4 +1110,268 @@ private boolean checkTraits(boolean currentExpectedToBeWorking) {
+        final String symbolicName = "test.duplicate.type."+JavaClassNames.niceClassAndMethod();
+        final String entityName = symbolicName+".type";
+        final String entityNameOkay = symbolicName+".okayType";
+
+        File jar = createJar(ImmutableMap.<String, String>of("catalog.bom", Joiner.on("\n").join(
+                "brooklyn.catalog:",
+                "  bundle: " + symbolicName,
+                "  version: " + TEST_VERSION,
+                "  items:",
+                "  - id: " + entityNameOkay,
+                "    itemType: entity",
+                "    item:",
+                "      type: org.apache.brooklyn.core.test.entity.TestEntity",
+                "  - id: " + entityName,
+                "    itemType: entity",
+                "    item:",
+                "      type: org.apache.brooklyn.core.test.entity.TestEntity",
+                "      a.config: first_definition",
+                "  - id: " + entityName,
+                "    itemType: entity",
+                "    item:",
+                "      type: org.apache.brooklyn.core.test.entity.TestEntity",
+                "      a.config: second_definition_makes_it_different_so_disallowed")));
+
+        Response result = client().path("/catalog/bundles")
+                .header(HttpHeaders.CONTENT_TYPE, "application/x-jar")
+                .post(Streams.readFully(new FileInputStream(jar)));
+        HttpAsserts.assertNotHealthyStatusCode(result.getStatus());
+        String resultS = Streams.readFullyString((InputStream)result.getEntity());
+        Asserts.assertStringContainsIgnoreCase(resultS, "different plan", entityName);
+        Asserts.assertStringDoesNotContain(resultS, entityNameOkay);
+
+        // entity not added
+        Response get = client().path("/catalog/types/" + entityName + "/" + TEST_VERSION).get();
+        assertEquals(get.getStatus(), 404);
+        
+        List<TypeSummary> entities = client().path("/catalog/types/" + entityName)
+                .get(new GenericType<List<TypeSummary>>() {});
+        Asserts.assertSize(entities, 0);
+        
+        // nor is the okay entity
+        Response getOkay = client().path("/catalog/types/" + entityNameOkay + "/" + TEST_VERSION).get();
+        assertEquals(getOkay.getStatus(), 404);
+        
+        // and nor is the bundle
+        Response getBundle = client().path("/catalog/bundles/" + symbolicName + "/" + TEST_VERSION).get();
+        assertEquals(getBundle.getStatus(), 404);
+    }
+    
+    // TODO might in future want to allow this if the user adding the type cannot see the other type due to entitlements;
+    // means however there might be another user who _can_ see the two different types 
+    private final static boolean DISALLOW_DIFFERENCES_IN_SAME_TYPE_ID_FROM_DIFFERENT_BUNDLES = true;
+    
+    @Test
+    public void testAddSameTypeTwiceInDifferentBundleDifferentDefinition_Disallowed() throws Exception {
```

Comment:
I don't understand why this case is not allowed. Could you elaborate on this please @ahgittin ?

Reply:
it might be allowed in future, it certainly can be behind the scenes, but it can cause oddness as it is then ambiguous what `type:version` would mean if there are two different implementations.  so there is strong argument to discourage/prevent it.

however in a multi-tenant world different users might not be allowed to see other bundles, so user 1 and user 2 might add the same `type:version` where they aren't allowed to see the other (and in fact shouldn't be allowed to know that someone else has defined that `type:version`), so in that case we _would_ allow this.  (we'd weaken restrictions to allow it, maybe e.g. user 2 can add `type:version` so long as he can't see any differing `type:version`.)  _but_ it might mean that a user 3 who is allowed to see bundles from both user 1 and user 2 is then able to see differing definition, which is not ideal and potentially confusing but i think least bad option.  cc @aledsage ?

(NB longer term discussion, not part of this PR which just clarifies and tests current behaviour in this regard)

Reply:
Ah got it, makes sense